src/main/scala/za/co/absa/spark_partition_sizing/DataFramePartitioner.scala

```diff
@@ -56,26 +56,39 @@ object DataFramePartitioner {
 
     def repartitionByPlanSize(minPartitionSize: Option[ByteSize], maxPartitionSize: Option[ByteSize]): DataFrame = {
 
-      def computeBlockCount(totalByteSize: BigInt, desiredSize: Long, addRemainder: Boolean): Int = {
+      def computeBlockCount(totalByteSize: BigInt, desiredSize: BigInt, addRemainder: Boolean): Int = {
         val int = (totalByteSize / desiredSize).toInt
         val blockCount = int + (if (addRemainder && (totalByteSize % desiredSize != 0)) 1 else 0)
         blockCount max 1
       }
 
+      def changePartitionCount(blockCount: Int, fnc: Int => DataFrame): DataFrame = {
+        val outputDf = fnc(blockCount)
+        outputDf
+      }
+
       df.cacheIfNot()
-      val catalystPlan = df.queryExecution.logical
-      val sizeInBytes = df.sparkSession.sessionState.executePlan(catalystPlan).optimizedPlan.stats.sizeInBytes
 
-      val currentBlockSize = sizeInBytes / df.rdd.getNumPartitions
+      val currentPartionCount = df.rdd.getNumPartitions
+
+      if (currentPartionCount > 0) {
+        val catalystPlan = df.queryExecution.logical
+        val sizeInBytes = df.sparkSession.sessionState.executePlan(catalystPlan).optimizedPlan.stats.sizeInBytes
 
-      (minPartitionSize, maxPartitionSize) match {
-        case (Some(min), None) if currentBlockSize < min =>
-          df.repartition(computeBlockCount(sizeInBytes, min, addRemainder = false))
-        case (None, Some(max)) if currentBlockSize > max =>
-          df.repartition(computeBlockCount(sizeInBytes, max, addRemainder = true))
-        case (Some(min), Some(max)) if currentBlockSize < min || currentBlockSize > max =>
-          df.repartition(computeBlockCount(sizeInBytes, max, addRemainder = true))
-        case _ => df
+        val currentBlockSize = sizeInBytes / df.rdd.getNumPartitions
+
+        (minPartitionSize, maxPartitionSize) match {
+          case (Some(min), None) if currentBlockSize < min =>
+            changePartitionCount(computeBlockCount(min, sizeInBytes, addRemainder = false), df.coalesce)
```

Comment:
Using `coalesce` here is smart 👍

Reply:
Still probably debateable 😅

Reply:
I actually think it's not desired. (I deliberately put `repartition` here.)
AFAIK `coalesce` just "put's together" some partitions to reach the desired number. I think the gola of the function in general should be to create roughly uniform partitions (unless there will be some special paritionings in the future). And with `coalesce` it's not gonna happen unless some further analysis is done (might be worth for those cases).

Reply:
Definitely debatable. Well, `coalesce` can result in _roughly_ uniform partitions, it's just that this _roughly_ may or may not be good enough. `repartition` will yield similar or better uniformity in exchange for more work. 🤷